src/core/qgsvectorlayerjoinbuffer.cpp

```diff
@@ -392,6 +394,44 @@ const QgsVectorLayerJoinInfo *QgsVectorLayerJoinBuffer::joinForFieldIndex( int i
   return &( mVectorJoins[sourceJoinIndex] );
 }
 
+QList<const QgsVectorLayerJoinInfo *> QgsVectorLayerJoinBuffer::joinsWhereFieldIsId( const QgsField &field ) const
+{
+  QList<const QgsVectorLayerJoinInfo *> infos;
+
+  for ( int i = 0; i < mVectorJoins.count(); i++ )
+  {
+    const QgsVectorLayerJoinInfo *info = &( mVectorJoins[i] );
+
+    if ( infos.contains( info ) )
+      continue;
+
+    if ( info->targetFieldName() == field.name() )
+      infos.append( info );
+  }
+
+  return infos;
+}
+
+QgsFeature QgsVectorLayerJoinBuffer::joinedFeatureOf( const QgsVectorLayerJoinInfo &info, const QgsFeature &feature ) const
+{
+  QgsFeature joinedFeature;
+
+  if ( info.joinLayer() )
+  {
+    const QVariant targetValue = feature.attribute( info.targetFieldName() );
+    const QString filter = QString( "\"%1\" = %2" ).arg( info.joinFieldName(), targetValue.toString() );
```

Comment:
Better use the QgsExpression quotedColumnRef/quotedValue methods here. That'll safely handle more edge cases.